pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,300 @@
+	ResourceNames []string
+	// NonResourceURLs is a set of partial urls that a user should have access to.  *s are allowed, but only as the full, final step in the path
+	// If an action is not a resource API request, then the URL is split on '/' and is checked against the NonResourceURLs to look for a match.
+	// Since non-resource URLs are not namespaced, this field is only applicable for ClusterRoles referenced from a ClusterRoleBinding.
+	// Rules can either apply to API resources (such as "pods" or "secrets") or non-resource URL paths (such as "/api"),  but not both.
+	// +optional
+	NonResourceURLs []string
+}
+
+// FlowSchemaStatus represents the current state of a flow-schema.
+type FlowSchemaStatus struct {
+	// Current state of flow-schema.
+	Conditions []FlowSchemaStatusCondition
+}
+
+// FlowSchemaStatusCondition describes conditions for a flow-schema.
+type FlowSchemaStatusCondition struct {
+	// Type is the type of the condition.
+	Type FlowSchemaConditionType
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status FlowSchemaConditionStatus
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	// Unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string
+	// Human-readable message indicating details about last transition.
+	Message string
+}
+
+// FlowSchemaStatusConditionType is a valid value for FlowSchemaStatusCondition.Type
+type FlowSchemaConditionType string
+
+// FlowSchemaStatusConditionStatus is a valid value for FlowSchemaStatusCondition.Status
+type FlowSchemaConditionStatus string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec
+	// Current status of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+	// Items is a list of request-priorities.
+	Items []PriorityLevelConfiguration
+}
+
+type PriorityLevelConfigurationSpec struct {
```

Comment:
This struct also needs a `CatchAll bool` field, to identify the prority level that is the default.  See https://github.com/MikeSpreitzer/kube-enhancements/blob/add-req-fq-isolated/keps/sig-api-machinery/20190228-priority-and-fairness.md#default-behavior .

Reply:
> This struct also needs a CatchAll bool field, 

i'm trying to eliminate the `CatchAll` field and convert it to a matching rule that applied to all users and all resources. do we really need that?

Reply:
I am not sure I understand the alternative being proposed.  Can you explain it?

My thinking here is that we need to define what happens to a request that is not matched by any FlowSchema.  My proposed answer is to assign it to one of two priority levels, depending on whether the request should be exempt or not.  In the "not" case, we need to know which priority level gets used.  And I am trying to define default behavior in a way that integrates smoothly with regular behavior, so disruption is minimized as these API objects are created/updated/deleted.

Reply:
so there should be at most 1 `catchAll` priority globally?

Reply:
Normally, yes.  But during transitions there might be any number.  That is why I prescribed the behavior in such a way that it is reasonable regardless of the number that exists at the moment.

Reply:
@MikeSpreitzer i see a similar field in scheduling api 

https://github.com/kubernetes/kubernetes/blob/c4c9e256c61cfed0eee34750fed3e61d69912080/pkg/apis/scheduling/types.go#L53-L59

how do you think if we if make the field name consistent ?

Reply:
OK with me.